backend/test/unit/openai.test.ts

```diff
@@ -69,6 +77,123 @@ describe('openAI unit tests', () => {
 		expect(validModels).toEqual(expectedValidModels);
 	});
 });
-afterEach(() => {
-	jest.clearAllMocks();
+
+const mockSystemRolePrompt =
+	'You are a helpful chatbot that answers questions.';
+const mockDefencesSystemRoleInactive: Defence[] = [
+	{
+		id: DEFENCE_ID.SYSTEM_ROLE,
+		config: [
+			{
+				id: 'SYSTEM_ROLE',
+				value: mockSystemRolePrompt,
+			},
+		],
+		isActive: false,
+		isTriggered: false,
+	},
+];
+const mockDefencesSystemRoleActive = [
+	{ ...mockDefencesSystemRoleInactive[0], isActive: true },
+];
+const mockChatHistoryWithoutSystemRole: ChatHistoryMessage[] = [
+	{
+		completion: { role: 'user', content: 'What is two plus two?' },
+		chatMessageType: CHAT_MESSAGE_TYPE.USER,
+	},
+	{
+		completion: { role: 'assistant', content: 'Two plus two equals four.' },
+		chatMessageType: CHAT_MESSAGE_TYPE.BOT,
+	},
+];
+
+const mockChatHistoryWithSystemRole: ChatHistoryMessage[] = [
+	{
+		completion: { role: 'system', content: mockSystemRolePrompt },
+		chatMessageType: CHAT_MESSAGE_TYPE.SYSTEM,
+	},
+	...mockChatHistoryWithoutSystemRole,
+];
+
+const mockIsDefenceActive = jest.fn();
+
+jest.mock('@src/defence', () => ({
+	getSystemRole: jest.fn().mockImplementation(() => mockSystemRolePrompt),
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-return
+	isDefenceActive: jest.fn().mockImplementation(() => mockIsDefenceActive()),
+}));
```

Comment:
I find this test suite a bit confusing. We have two top-level describe blocks, sharing some setup/teardown, but then we have a bunch of consts and mocks which are also at the top-level but only used in one or other of the describes.

This might be a symptom of not knowing where some of this logic should sit. If there are functions in the module under test that only use a subset of the imports, then it's ok to stub them at the top level, but then you should mock the behaviour only where it is needed. So here, for example, we wouldn't necessarily want to `mockImplementation` for getSystemRole for tests within a describe that don't actually call getSystemRole.

And finally, if you are mocking the `@src/defence` module, then I believe getSystemRole and isDefenceActive are already empty mocks:

```
import { isDefenceActive } from '@src/defence';

jest.mock('@src/defence');

describe('some function', () => {
  const isDefenceActiveMock = isDefenceActive as jest.Mocked<typeof isDefenceActive>;

  afterEach(() => {
    isDefenceActiveMock.mockReset();
  });

  test('something', () => {
    isDefenceActiveMock.mockImplementationOnce(() => true);
    ...
  });
});
```

More info at https://jestjs.io/docs/mock-function-api#jestmockedsource